charms/tensorboards-web-app/src/charm.py

```diff
@@ -1,158 +1,264 @@
+        """Create and return Pebble framework layer."""
+        exec_command = (
+            "gunicorn" " -w 3" f" --bind 0.0.0.0:{PORT}" " --access-logfile" " - entrypoint:app"
+        )
+
+        layer_config = {
+            "summary": "jupyter-web-app layer",
+            "description": "Pebble config layer for jupyter-web-app",
+            "services": {
+                self._container_name: {
+                    "override": "replace",
+                    "summary": "Entrypoint of jupyter-web-app image",
+                    "command": exec_command,
+                    "startup": "enabled",
+                    "environment": self._env_vars,
+                    "on-check-failure": {"tensorboards-web-app-up": "restart"},
+                }
+            },
+            "checks": {
+                "tensorboards-web-app-up": {
+                    "override": "replace",
+                    "period": "30s",
+                    "http": {"url": f"http://localhost:{PORT}"},
                 },
-                "containers": [
-                    {
-                        "name": "tensorboards-web-app",
-                        "imageDetails": image_details,
-                        "envConfig": {
-                            "USERID_HEADER": "kubeflow-userid",
-                            "USERID_PREFIX": "",
-                            "APP_PREFIX": "/tensorboards",
-                            "APP_SECURE_COOKIES": str(config["secure-cookies"]),
-                            "BACKEND_MODE": config["backend-mode"],
-                        },
-                        "ports": [{"name": "http", "containerPort": config["port"]}],
-                    }
-                ],
             },
-        )
+        }
+
+        return Layer(layer_config)
+
+    def _on_install(self, _) -> None:
+        """Perform installation only actions."""
+        try:
+            # Deploy K8S resources to speed up deployment
+            self._deploy_k8s_resources()
+        except ErrorWithStatus as err:
+            self._log_and_set_status(err.status)
+        return
+
+    def _on_remove(self, _) -> None:
+        """Remove all resources."""
+        self._log_and_set_status(MaintenanceStatus("Removing K8S resources"))
+        k8s_resources_manifests = self.k8s_resource_handler.render_manifests()
+        try:
+            delete_many(self.k8s_resource_handler.lightkube_client, k8s_resources_manifests)
+        except ApiError as err:
+            # Do not log/report when resources were not found
+            if err.status.code != 404:
+                self.logger.error(f"Removing K8S resources failed with error: {err}")
+                raise GenericCharmRuntimeError("Removing K8s resources failed") from err
+
+    def _on_event(self, event) -> None:
+        """Perform required actions for every event."""
+        try:
+            self._check_leader()
+            self._deploy_k8s_resources()
+            update_layer(
```

Comment:
Not sure if you missed it when pushing new changes, but we definitely want to [check the container connection](https://github.com/canonical/kubeflow-tensorboards-operator/blob/main/charms/tensorboard-controller/src/charm.py#L241) before attempting to update the Pebble layer.

Reply:
I had this comment https://github.com/canonical/kubeflow-tensorboards-operator/pull/74#discussion_r1237207799. I resolved the conversation after @i-chvets's response but maybe I shouldn't have. Isn't [this check](https://github.com/canonical/charmed-kubeflow-chisme/blob/c82bd1dd21b375c6cf91f43298fdc09d98e4e07c/src/charmed_kubeflow_chisme/pebble/_update_layer.py#L22-L24) enough? @DnPlas

Reply:
Should we do this before also attempting to `_deploy_k8s_resources()`?